In [27]:
import numpy as np

In [28]:
def find_nearest_both(patern, codes, vectors):
    result = []
    for i in range(len(vectors)):
        l1_d = np.linalg.norm(vectors[i] - patern, 1)
        l2_d = np.linalg.norm(vectors[i] - patern)
        result.append([codes[i], (l1_d, l2_d), l1_d, l2_d])
    return sorted(result,key=lambda x: x[1],reverse=False)

def find_nearest(patern, codes, vectors, distance):
    result = []
    for i in range(len(vectors)):
        d = 0
        if distance == 1:
            d = np.linalg.norm(vectors[i] - patern, 1)
        if distance == 2:
            d = np.linalg.norm(vectors[i] - patern)
        result.append([codes[i], d])
    return sorted(result,key=lambda x: x[1],reverse=False)

debuts = np.loadtxt("debuts.csv", quotechar='"', delimiter=',', dtype=str)
debuts_data = {}
for d in debuts[1:]:
    debuts_data[d[0]] = d[1]
chess_openings = np.loadtxt("chess_openings.csv", delimiter=",", dtype=str)
code = 'A54'
name = 'Old Indian, Ukrainian Variation, 4.Nf3 (A54)'
codes = []
vectors = []
patern = []
for i in chess_openings:
    c = i[0]
    l = np.array(list(map(float, i[1:])))
    if code == c:
        patern = l
    else:
        codes.append(c)
        vectors.append(l)


result_l1 = find_nearest(patern, codes, vectors, 1)[:10]
result_l2 = find_nearest(patern, codes, vectors, 2)[:10]
result_both = find_nearest_both(patern, codes, vectors)[:10]

print("Nearest L1")
for n in range(10): print(result_l1[n], debuts_data[result_l1[n][0]])

print("Nearest L2")
for n in range(10): print(result_l2[n], debuts_data[result_l2[n][0]])

print("Diff")
l1r = [k for k,d in result_l1]
l2r = [k for k,d in result_l2]
print(set(l1r) - set(l2r))

Nearest L1
['A55', 3.46] Old Indian, Main line (A55)
['A53', 4.183999999999999] Old Indian (A53)
['C41', 5.821] Philidor Defense (C41)
['A21', 7.284000000000001] English (A21)
['E61', 7.336] King's Indian (E61)
['A41', 7.402] Queen's Pawn Game (with ...d6) (A41)
['A56', 7.513] Benoni Defense (A56)
['E95', 7.605] King's Indian, Orthodox, 7...Nbd7, 8.Re1 (E95)
['A27', 7.661] English, Three Knights System (A27)
['E96', 7.755] King's Indian, Orthodox, 7...Nbd7, Main line (E96)
Nearest L2
['A53', 0.8844648099274496] Old Indian (A53)
['A55', 1.0106176329354244] Old Indian, Main line (A55)
['A41', 1.5177107761362174] Queen's Pawn Game (with ...d6) (A41)
['A21', 1.5303130398712537] English (A21)
['C41', 1.632522281624359] Philidor Defense (C41)
['A51', 1.767700766532616] Budapest Gambit (A51)
['A50', 1.7723806024666366] Queen's Pawn Game (A50)
['A28', 1.7962488691715297] English (A28)
['A27', 1.805460882988053] English, Three Knights System (A27)
['A20', 1.8558246145581747] English (A20)
Diff



Nearest L1
['A55', 3.46] Old Indian, Main line (A55)
['A53', 4.183999999999999] Old Indian (A53)
['C41', 5.821] Philidor Defense (C41)
['A21', 7.284000000000001] English (A21)
['E61', 7.336] King's Indian (E61)
['A41', 7.402] Queen's Pawn Game (with ...d6) (A41)
['A56', 7.513] Benoni Defense (A56)
['E95', 7.605] King's Indian, Orthodox, 7...Nbd7, 8.Re1 (E95)
['A27', 7.661] English, Three Knights System (A27)
['E96', 7.755] King's Indian, Orthodox, 7...Nbd7, Main line (E96)

Nearest L2
['A53', 0.8844648099274496] Old Indian (A53)
['A55', 1.0106176329354244] Old Indian, Main line (A55)
['A41', 1.5177107761362174] Queen's Pawn Game (with ...d6) (A41)
['A21', 1.5303130398712537] English (A21)
['C41', 1.632522281624359] Philidor Defense (C41)
['A51', 1.767700766532616] Budapest Gambit (A51)
['A50', 1.7723806024666366] Queen's Pawn Game (A50)
['A28', 1.7962488691715297] English (A28)
['A27', 1.805460882988053] English, Three Knights System (A27)
['A20', 1.8558246145581747] English (A20)